tracer/src/Datadog.Trace/Telemetry/Collectors/DependencyTelemetryCollector.cs

```diff
@@ -71,12 +69,23 @@ public bool HasChanges()
         public ICollection<DependencyTelemetryData> GetData()
         {
             var hasChanges = Interlocked.CompareExchange(ref _hasChangesFlag, 0, 1) == 1;
+
             if (!hasChanges)
             {
                 return null;
             }
 
-            return _assemblies.Keys;
+            var assembliesToRecord = new List<DependencyTelemetryData>();
+            foreach (var assembly in _assemblies)
+            {
+                if (assembly.Value == false)
+                {
+                    _assemblies[assembly.Key] = true;
+                    assembliesToRecord.Add(assembly.Key);
+                }
+            }
```

Comment:
Just spotted an issue with this. The code in https://github.com/DataDog/dd-trace-dotnet/blob/4d6154ce137f9bac30703659c3ac03626f1b612a/tracer/src/Datadog.Trace/Telemetry/TelemetryController.cs#L216 assumes that `GetData()` returns _all_ the assemblies, so it can tolerate transport errors.

With this new approach, you could have the following:

- Initial `GetData()` returns 50 assemblies
- Sending fails. `TelemetryController` remembers failed list
- A new assembly is referenced
- Next clock, `TelemetryController` calls `GetData()`. 
- This returns the _single_ assembly. `TelemetryController` throws away the stale data.
- We never record the original 50 assemblies

So something has to change 🙁 

One possible option - keep this code the same, but change [this line](https://github.com/DataDog/dd-trace-dotnet/blob/4d6154ce137f9bac30703659c3ac03626f1b612a/tracer/src/Datadog.Trace/Telemetry/TelemetryController.cs#L216) to something like this:

```csharp
var dependencies = _dependencies.GetData();
if(dependencies is null)
{
    dependencies = _circuitBreaker.PreviousDependencies;
}
else if(_circuitBreaker.PreviousDependencies is {} oldDeps)
{
    dependencies.AddRange(oldDeps);
}
```

Reply:
Thanks a lot for spotting this and for the suggestion.
Done in 66e8c76e02d6299365cd25bb43bfb4dd2d4ebad2